src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -348,13 +432,62 @@ public void DoDisconnectIfNeeded()
                 }
 
                 Self._socket.Dispose();
+                Self.ReturnAllBlocks();
                 Self._log.ConnectionStop(Self._connectionId);
                 Complete();
             }
 
             public void Complete()
             {
-                Self.OnWriteCompleted(Buffers, WriteStatus, WriteError);
+                Self.OnWriteCompleted(_byteCount, WriteStatus, WriteError);
+            }
+
+            private void ReturnFullyWrittenBlocks()
+            {
+                var block = _lockedStart.Block;
+                while (block != _lockedEnd.Block)
+                {
+                    var returnBlock = block;
+                    block = block.Next;
+
+                    returnBlock.Unpin();
+                    returnBlock.Pool?.Return(returnBlock);
+                }
+            }
+
+            private void LockWrite()
```

Comment:
Not sure how this inert the blocks currently passed out on ProducingStart as part of the MemoryPoolIterator2